librz/core/cmd_zign.c

```diff
@@ -674,6 +675,21 @@ static int searchHitCB(RzSignItem *it, RzSearchKeyword *kw, ut64 addr, void *use
 	apply_flag (ctx->core, it, addr, kw->keyword_length, kw->count, ctx->prefix, ctx->rad);
 	RzAnalysisFunction *fcn = rz_analysis_get_fcn_in (ctx->core->analysis, addr, 0);
 	// TODO: create fcn if it does not exist
+	if (fcn && ctx->wrapper_bb && ctx->wrapper_bb->jump == addr) {
+		ut64 wrapper_size = ctx->wrapper_bb->size;
+		ctx->wrapper_bb = NULL;
+		RzAnalysisFunction *newfcn = rz_analysis_create_function (ctx->core->analysis, NULL, addr, RZ_ANALYSIS_FCN_TYPE_FCN, NULL);
+		ut64 size = rz_analysis_function_realsize (fcn) - wrapper_size;
+		int flen = rz_analysis_fcn (ctx->core->analysis, newfcn, addr, size, RZ_ANALYSIS_REF_TYPE_CALL);
+		/* rz_analysis_fcn() returns RZ_ANALYSIS_RET_ERROR if BBs already exist */
+		if (flen == RZ_ANALYSIS_RET_ERROR) {
```

Comment:
@thestr4ng3r maybe you know a better way to split a function

Reply:
I think it's very questionable whether the zignature search should modify functions and basic blocks at all. Especially `rz_analysis_fcn()` should be avoided at all costs since it is the actual analysis loop that could mess up a lot of things. Are there any other cases where more than just adding a flag, renaming and maybe applying types happens? Otherwise I would suggest to just add a flag in the middle of the function and leave it at that.

Anyway, in case we do decide to split functions here, you first have to be aware that `rz_analysis_get_fcn_in()` used above is a very fuzzy and half-deprecated thing. Since both functions and basic blocks can overlap, considering a single function containing `addr` here is conceptually wrong. Intuitively I would say `rz_analysis_get_function_at()` should be used above and for your wrapped functions, you will want to save the wrapper function next to the `wrapper_bb`.

Once you really do have the correct function that at this point contains the wrapper bb and the rest of the function, as well as a reference to this wrapper bb, you can call `rz_analysis_function_relocate()` on the function and set its entrypoint to the target of the wrapper bb. Then you can remove the block from the function with `rz_analysis_function_remove_block()`, create a new one for the wrapper function with `rz_analysis_create_function()` and add the bb to that one with `rz_analysis_function_add_block()`. Basically the other way around as you currently did it.
Also, be aware of the refcounting of the blocks! If you didn't explicitly call `rz_analysis_block_ref()` and you didn't get the block reference out of a function that did this for you, after removing the block from the last function, it would be freed automatically!

Reply:
@thestr4ng3r In future I could see zignature also containing color highlights of BBs, comments, calling convention, signatures, local flags, and much more that, could really be handy imho.

Reply:
Perhaps it should be a two-step process in some way. First searching and then optionally applying to functions.

Reply:
I understand that my code can break some internal data. That's why I have made draft PR. Anyway now I think it was bad idea to sent it. It would be better to implement a complete solution, not an interim one.